SAT/Z3Interfacing.cpp

```diff
@@ -1145,9 +1154,44 @@ Z3Interfacing::~Z3Interfacing()
   _sorts.clear();
   _toZ3.clear();
   _fromZ3.clear();
+  outputln(); // flush the output file
+}
+
+
+
+bool Z3Interfacing::isNamedExpr(unsigned var) const
+{ return _varNames.find(var); }
+
+z3::expr Z3Interfacing::getNameExpr(unsigned var)
+{
+  return _varNames.getOrInit(var, [&](){
+      // this method is called very often in runs with a lot of avatar reasoning. Cache the constants to avoid that z3 has to search for the string name in its function index
+      vstring name = "v"+Lib::Int::toString(var);
+      outputln("(declare-fun ", name, " () Bool)");
+      return _context.bool_const(name.c_str());
+  });
 }
 
 
+z3::expr Z3Interfacing::getNamingConstantFor(TermList toName, z3::sort sort)
+{
+  return _termIndexedConstants.getOrInit(toName, [&](){
+      auto name = "n" + toName.toString();
```

Comment:
We never actually looked at whether this single letter prefix was enough to heuristically avoid clashes, which could lead to unsoundness (I think)... the same for the new prefix for constants below. There's obviously a way we could do it